### PR TITLE
Add blogpost for new KinesisStreamsSource and DynamoDbStreamsSource

### PR DESCRIPTION
**Pull request by hlteoh37**
Add a blogpost to illustrate in detail the KDS and DDB streams source

Image of blogpost included: 
KDS_DB_blog

### CODE REVIEW

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+</table>
+
+
+## Why did we need new source connectors?
+
+We implemented new source connectors because the legacy `FlinkKinesisConsumer` and `FlinkDynamoDBStreamsConsumer` use the legacy `SourceFunction` interface, which is removed in Flink 2.x. As such, from Flink 2.x onwards, only sources using the new `Source` interface will be supported.
+
+## New features
+
+Moving over to the new `Source` interface introduced the following advantages:
```
Comment (nicusX) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: Something wrong in this sentence. 
Maybe "The new `Source` interface introduced..." would suffice

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+- **Native integration with Flink watermarks.** On the new `Source` interface, watermark generation is abstracted away to the Flink framework, and no longer a responsibility of the source. This means the new source has support for watermark alignment, and idle watermark handling out-of-the-box.
+- **Standardised Flink Source metrics.** The new `Source` framework also introduces [standardised Source metrics](https://cwiki.apache.org/confluence/display/FLINK/FLIP-33%3A+Standardize+Connector+Metrics). This means that tracking record throughput and lag can be easily tracked.
```
Comment (nicusX) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: "tracking" and "tracked" are redundant
"This means that record throughout and lag can be easily tracked"

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+- **Records are read in-order even after a resharding operation on the stream.** The new sources ensure that parent shards are read completely before reading children shards. This allows record ordering to be maintained even after a resharding operation. See [Detailed explanation of record ordering](#appendix-detailed-explanation-of-record-ordering) for more information.
```
Comment (nicusX) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: "The new **source**" (not "sources")

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+We also take this opportunity to perform the following improvements to the connector:
```
Comment (nicusX) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: "...**took** the opportunity..."?

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+- **Migrate away from AWS SDK v1 to AWS SDK v2.** This SDK update aligns with best practices.
```
Comment (nicusX) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: "best practices" of whom? Maybe better saying it has improved performance, stability etc

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+- **Migrate away from custom retry strategies to use the AWS SDK native retry strategies instead.** This allows us to benefit from AWS error classification as well as retry algorithm.
+- **Reduce jar size by >99%, from ~60MB to ~100KB.** In the new source, we no longer shade the AWS SDK and no longer need to package the Kinesis Producer Library (needed for legacy sink). This should lead to smaller Flink application jars.
+- **Improve defaults.** For the `FlinkKinesisConsumer`, we implemented an improved `UniformShardAssigner` which provides even shard distribution across subtasks even when there is a resharding event, by using information around the Shard hash key range. This was not made default as it would have been a breaking change. In the new `KinesisStreamsSource`, this is now the default shard assigner.
```
Comment (nicusX) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: This is a bit confusing.
I would say simply something like
"`UniformShardAssigner` which provides even shard distribution across subtasks even when there is a resharding event, by using information around the Shard hash key range, is now the default assigner in the new `KinesisStreamsSource`"

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+During the implementation of the source Table API, we had to make some breaking changes around the table identifier of `kinesis`. As such, we had to do a major version bump from `4.x` to `5.x`. In Table API / SQL, for version `4.x` and below, `kinesis` refers to the `FlinkKinesisConsumer`. However, from `5.x` onwards, `kinesis` now refers to the `KinesisStreamSource`. To use the legacy `FlinkKinesisConsumer`, you can use the table identifier of `kinesis-legacy`. See [`Kinesis Table API documentation`](https://nightlies.apache.org/flink/flink-docs-stable/docs/connectors/table/kinesis/) for more details.
```
Comment (nicusX) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: "...we had to introduce some breaking changes..." [...] "This is the reason we have a major version bump [or change], from `4.x` to `5.x`."

Comment (nicusX) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: The connector naming is hard to follow.
Maybe something like:
"Because of the new connector interface, the connector's Table API / SQL name has also changed. In version `4.x`, source connector `kinesis` refers to `FlinkKinesisConsumer`. 
However, from `5.x` onwards, `kinesis` refers to the new `KinesisStreamSource`. 
You can still use the old `FlinkKinesisConsumer` connector with `5.x` using the identifier `kinesis-legacy`"

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+## Migration guidance
+
+There is no state compatibility between the legacy sources (`FlinkKinesisConsumer` and `FlinkDynamoDBStreamsConsumer`), and the new sources (`KinesisStreamsSource` and `DynamoDbStreamsSource`). This means that in order to migrate from the legacy source to the new source, users must:
```
Comment (nicusX) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: "... the user must reset the state of the source operator and start from a known position, back in the stream, to prevent any data loss."
"If you are following the best practices of setting the uuid of all operators you can  do the following:"

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+1. Change the uuid of the source operator.
+2. Ensure that when restoring from a savepoint, non-restored state is allowed.
+
+To attempt to pick up where the old source left off, users can start the new `KinesisStreamsSource` from a specified timestamp that is slightly in the past, to maintain at least once processing of records. This can be done by specifying `KinesisSourceConfigOptions.STREAM_INITIAL_POSITION` as `AT_TIMESTAMP` and specifying the preferred timestamp with `KinesisSourceConfigOptions.STREAM_INITIAL_TIMESTAMP`.
```
Comment (nicusX) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: "To prevent data loss, users can start..."

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+## Summary
+The `KinesisStreamsSource` and `DynamoDbStreamsSource` deliver much needed new features for the community to use! Do feel free to reach out on the Flink mailing list (dev@flink.apache.org) or Flink Slack to discuss any further improvements.
+
+## Further work
+There are some future developments in the plans:
+- Support for Table API and SQL for the DynamoDB Streams Source. Ideally, we also integrate with CDC so that changes made to the DynamoDB table are understood natively in Flink's Table API environment.
+- Python integration for both sources are not implemented yet, but would be useful to have!
```
Comment (nicusX) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: "**DataStream** API Python integration..."

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 1 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: Set up as 
```
The new Source interface introduces:
```

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 2 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: Done

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 3 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: Done

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 4 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: Changed to - "We also introduce the following improvements to the connector"

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 5 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: Technically it doesnt 👀 I couldn't find anything at least!

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 6 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: done

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 7 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: done

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 10 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: done

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 9 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: done

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 8 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: done

**Review by nicusX — APPROVED** [review]
LGTM

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+      <a href="https://nightlies.apache.org/flink/flink-docs-stable/docs/connectors/datastream/dynamodb/">
+      Flink DynamoDB connector documentation</a> for more details.
+    </td>
+  </tr>
+</table>
+
+
+## Why did we need new source connectors?
+
+We implemented new source connectors because the legacy `FlinkKinesisConsumer` and `FlinkDynamoDBStreamsConsumer` use the deprecated `SourceFunction` interface, which is removed in Flink 2.x. From Flink 2.x onwards, only `KinesisStreamsSource` and `DynamoDbStreamsSource`, which use the new `Source` interface will be supported.
```
Comment (dannycranmer) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]:
```suggestion
We implemented new source connectors because the `FlinkKinesisConsumer` and `FlinkDynamoDBStreamsConsumer` use the deprecated `SourceFunction` interface, which is removed in Flink 2.x. From Flink 2.x onwards, only `KinesisStreamsSource` and `DynamoDbStreamsSource`, which use the new `Source` interface will be supported.
```

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+      <a href="https://github.com/apache/flink-connector-aws/blob/v5.0/flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/source/DynamoDbStreamsSourceBuilder.java">
+      DynamoDbStreamsSourceBuilder</a> to create the source. See 
+      <a href="https://nightlies.apache.org/flink/flink-docs-stable/docs/connectors/datastream/dynamodb/">
+      Flink DynamoDB connector documentation</a> for more details.
+    </td>
+  </tr>
+</table>
+
+
+## Why did we need new source connectors?
```
Comment (dannycranmer) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md:64]: This section undersells it I think? Users also benefit from the improvements in the new Source API

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+2. **Standardised Flink Source metrics.** The new `Source` framework also introduces [standardised Source metrics](https://cwiki.apache.org/confluence/display/FLINK/FLIP-33%3A+Standardize+Connector+Metrics). This enables users to track record throughput and lag across sources in a standardised manner.
+3. **Records are read in-order even after a resharding operation on the stream.** The new `Source` ensures that parent shards are read completely before reading children shards. This allows record ordering to be maintained even after a resharding operation. See [explanation of record ordering in Kinesis Data Streams](#appendix-detailed-explanation-of-record-ordering) for more information.
+4. **Migrate away from AWS SDK v1 to AWS SDK v2.** This SDK update aligns with best practices.
+5. **Migrate away from custom retry strategies to use the AWS SDK native retry strategies instead.** This allows us to benefit from AWS error classification in the retry algorithm.
```
Comment (dannycranmer) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]:
```suggestion
5. **Migrate away from custom retry strategies to use the AWS SDK native retry strategies.** This allows us to benefit from AWS error classification in the retry algorithm.
```

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+6. **Reduce jar size by >99%, from [~60MB](https://mvnrepository.com/artifact/org.apache.flink/flink-connector-kinesis/5.0.0-1.20) to [~200KB](https://mvnrepository.com/artifact/org.apache.flink/flink-connector-aws-kinesis-streams/5.0.0-1.20).** In the new source, we no longer shade the AWS SDK and no longer need to package the Kinesis Producer Library (needed for legacy sink). This will lead to smaller Flink application jars.
```
Comment (dannycranmer) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: > This will lead to smaller Flink application jars.

However the customer will still need to shade the AWS SDK into the final application jar. Might want to make this more explicit incase it adds confusion

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+7. **Improve defaults.** The `UniformShardAssigner` which provides even shard distribution across subtasks even when there is a resharding event, by using information around the Shard hash key range, is now the default shard assigner.
```
Comment (dannycranmer) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: > even when there is a resharding event

Note: before this was always a problem, not just during resharding

Comment (dannycranmer) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]:
```suggestion
7. **Improve defaults.** The `UniformShardAssigner` is now the default shard assigner. This change results in a uniform shard distribution across Flink subtasks and can reduce unexpected processing skew 
```

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+
+## Migration guidance
+
+There is no state compatibility between the legacy sources (`FlinkKinesisConsumer` and `FlinkDynamoDBStreamsConsumer`), and the new sources (`KinesisStreamsSource` and `DynamoDbStreamsSource`). This means that in order to migrate from the legacy source to the new source, users must drop the state of the source operator and start from a specified starting position, to prevent any data loss. If users are following the best practice of setting uuid for all operators, the migration steps are the following:
+
+1. Change the uuid of the source operator.
+2. Ensure that when restoring from a savepoint, non-restored state is allowed.
+
+To prevent data loss, users can start the new `KinesisStreamsSource` from a specified timestamp that is slightly in the past, to maintain at least once processing of records. This can be done by specifying `KinesisSourceConfigOptions.STREAM_INITIAL_POSITION` as `AT_TIMESTAMP` and specifying the preferred timestamp with `KinesisSourceConfigOptions.STREAM_INITIAL_TIMESTAMP`.
```
Comment (dannycranmer) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: I wonder if we should just drop this all together? There are a lot of variables that could break things here and generate unexpected output, thoughts?

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+    <th>Usage</th>
+  </tr>
+  <tr>
+    <td>Amazon Kinesis Data Streams source</td>
+    <td>DataStream<br>Table API</td>
+    <td> Use the <code>flink-connector-aws-kinesis-streams</code> artifact. See <a href="https://nightlies.apache.org/flink/flink-docs-stable/docs/connectors/datastream/kinesis/">
+      Flink Kinesis connector documentation</a> for details.
+    </td>
+    <td>
+      Use the fluent
```
Comment (dannycranmer) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md:32]: It might be good to include a simple migration example here, showing how to migrate from Legacy to new

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+</table>
+
+
+## Why did we need new source connectors?
+
+We implemented new source connectors because the legacy `FlinkKinesisConsumer` and `FlinkDynamoDBStreamsConsumer` use the deprecated `SourceFunction` interface, which is removed in Flink 2.x. From Flink 2.x onwards, only `KinesisStreamsSource` and `DynamoDbStreamsSource`, which use the new `Source` interface will be supported.
+
+## New features
+
+The updated `KinesisStreamsSource` and `DynamoDbStreamsSource` connectors offer the following new features:
```
Comment (dannycranmer) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md:72]: How about feature parity (EFO) with the legacy connectors? Are there any gaps (the SDK v1 had more config options IIRC)

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+<center>
+<br/>
+<img src="/img/blog/2024-11-25-whats-new-aws-connectors/kinesis_records_sharding.png" width="60%"/>
+<br/>
+Fig. 2 - Illustration of record distribution within a Kinesis Data Stream after a resharding operation.
+</center>
+
+As we can see, to ensure that records from the `pk2` are read in order, we need to ensure that the shards are read in order of `Shard 0`, then `Shard 3`, then `Shard 6`. This can be more easily understood as : All parent shards must be fully read before children shard can be read.
+
+The new `KinesisStreamsSource` ensures that parent shards are read completely before reading the children shard, and so ensure that record ordering is maintained even after a resharding operation on the stream.
```
Comment (dannycranmer) [inline docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md:209]: Nice! 👏

**Review by dannycranmer — CHANGES_REQUESTED** [review]

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 11 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: adopted suggestion

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 12 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md:64]: Yep we cover it below. Will mention it here too.

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 13 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: accepted suggestion

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 14 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: Done

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 15 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: yes, but the other situation is hard to describe 👀 Effectively this is only a problem if there is resharding event. But it can manifest during live + resharding, or resharding + consume from backlog

Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: Adjusted

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 16 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md]: Happy to. It assumes too much IMO.

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 18 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md:72]: we have EFO support, but there are some future work still needed, mentioned in the further work section

File: docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md (hunk 17 above)
Comment (hlteoh37) [reply docs/content/posts/2024-11-25-whats-new-aws-connectors-5.0.0.md:32]: Done

**Comment by hlteoh37** [comment]
Updated snapshot
FireShot Capture 004 - Introducing the new Amazon Kinesis Data Stream and Amazon DynamoDB St_ - localhost

**Review by dannycranmer — APPROVED** [review]
